feat(core): add HOST_TAG_NAME token

devknoll opened the pull request:
For ideal DOM usage, you would not unwrap an ElementRef outside of the browser. However, it's reasonable to want to find the tag name of the host node on the server, and so this introduces a `HOST_TAG_NAME` token that can be injected to read this value.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

See #50621 for additional information.

eneajaho commented on packages/core/src/di/host_tag_name_token.ts:
```diff
@@ -0,0 +1,50 @@
+ * @publicApi
+ */
+export class HostTagNameToken {
+  constructor() {}
```
```suggestion
```

devknoll replied:
This is included as otherwise the constructor will not be documented in the public API

devknoll commented:
@AndrewKushnir

devknoll commented:
friendly ping @AndrewKushnir

AndrewKushnir commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -4507,6 +4507,93 @@ describe('di', () => {
+      }
+
+      expect(() => TestBed.createComponent(TestNgContainer))
+          .toThrowError(/No provider for HostTagNameToken found/);
```
I think it'd be great to improve this error message to make it more actionable. The `injectTagName` function would be invoked by the DI code with an argument that represents injection flags. If we are trying to inject a tag name on non-element nodes and there is no "optional" flag specified - we can throw an error and mention which host node is used (and why it doesn't have a tag name) and suggest to use the "optional" flag.

AndrewKushnir commented on packages/core/src/di/host_tag_name_token.ts:
````diff
@@ -0,0 +1,50 @@
+ * }
+ * ```
+ *
+ * ### Optionally injecting an a tag name
````
```suggestion
 * ### Optionally injecting a tag name
```

AndrewKushnir commented on packages/core/src/di/host_tag_name_token.ts:
```diff
@@ -0,0 +1,50 @@
+   */
+  __NG_ELEMENT_ID__: typeof injectTagName = injectTagName;
+
+  toString(): string {
+    return 'HostTagNameToken';
```
We can add an annotation to avoid exposing this method in docs (on angular.dev).
```suggestion
  /** @nodoc */
  toString(): string {
    return 'HostTagNameToken';
```

devknoll committed:
feat(core): add HOST_TAG_NAME token

For ideal DOM usage, you would not unwrap an ElementRef outside of the browser. However, it's reasonable to want to find the tag name of the host node on the server, and so this introduces a HOST_TAG_NAME token that can be injected to read this value.

AndrewKushnir commented on packages/core/src/di/host_tag_name_token.ts:
```diff
@@ -0,0 +1,55 @@
+// so we manually specify __NG_ELEMENT_ID__ here, instead of using a factory.
+// tslint:disable-next-line:no-toplevel-property-access
+(HOST_TAG_NAME as any).__NG_ELEMENT_ID__ = (flags: InjectFlags) => {
+  const tNode = getCurrentTNode()!;
+  if (tNode.type & TNodeType.Element) {
```
We can also add a check here to provide a better error message in case tag name is injected outside of the supported context:

```suggestion
  const tNode = getCurrentTNode();
  
  if (ngDevMode && tNode === null) {
    throw new RuntimeError(...);
  }
  
  if (tNode.type & TNodeType.Element) {
```

AndrewKushnir commented on packages/core/src/di/host_tag_name_token.ts:
```diff
@@ -0,0 +1,55 @@
+  throw new RuntimeError(
+      RuntimeErrorCode.INVALID_INJECTION_TOKEN,
+      ngDevMode &&
+          'HOST_TAG_NAME was used on a node without an underlying host node in the DOM. ' +
```
We can use `tNode.type` to determine if an underlying node was an `<ng-container>` or `<ng-template>` (and use "a node" as the default for other TNode types) and include that into the error message (the `node` part can be defined based on that):
```suggestion
          `HOST_TAG_NAME was used on ${node} which doesn't have an underlying element in the DOM. ` +
```

AndrewKushnir commented on packages/core/src/di/host_tag_name_token.ts:
```diff
@@ -0,0 +1,71 @@
+// so we manually specify __NG_ELEMENT_ID__ here, instead of using a factory.
+// tslint:disable-next-line:no-toplevel-property-access
+(HOST_TAG_NAME as any).__NG_ELEMENT_ID__ = (flags: InjectFlags) => {
+  const tNode = getCurrentTNode()!;
+  if (tNode === null) {
+    throw new RuntimeError(
+        RuntimeErrorCode.INVALID_INJECTION_TOKEN,
+        ngDevMode && 'HOST_TAG_NAME can only be used on directives and components');
+  }
```
```suggestion
  const tNode = getCurrentTNode();
  if (tNode === null) {
    throw new RuntimeError(
        RuntimeErrorCode.INVALID_INJECTION_TOKEN,
        ngDevMode && 'HOST_TAG_NAME can only be injected in directives and components ' +
          'during construction time (in a class constructor or as a class field initializer)');
  }
```

AndrewKushnir replied:
(note: we can also remove the `!` after the `getCurrentTNode()` call, since we check for `null` now)

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

thePunderWoman reviewed (approved):
reviewed-for: public-api

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api

alxhub commented:
This PR was merged into the repository by commit 5f06ca8f5539ed208bae0b110887b5538ac4041f.
